### PR TITLE
Intermediate tour: Functions chapter

### PR DESCRIPTION
**Pull request by sarahhaggarty**
This PR is to add the first chapter of the intermediate tour to the staging branch.

### CODE REVIEW

File: docs/topics/tour/kotlin-tour-intermediate-functions.md [hunk 1]
```diff
@@ -0,0 +1,935 @@
...
+```
+{initial-collapse-state="collapsed" collapsible="true" collapsed-title="Example solution" id="kotlin-tour-scope-functions-solution-2"}
+
+## Lambda expressions with receiver
+
+In the beginner's tour, you learned how to use [lambda expressions](kotlin-tour-functions.md#lambda-expressions). Lambda expressions can also have a receiver, 
+which means that the lambda expression can access any member functions or properties of the receiver object without having
+to explicitly specify it each time. Without these additional references, your code is easier to read and maintain.
+
+The syntax for a lambda expression with receiver is different when you define the function type: write the receiver object
```
Comment (danil-pavlov) [inline docs/topics/tour/kotlin-tour-intermediate-functions.md:656]: I found this a little hard to follow. Consider this:

> The syntax for a lambda expression with receiver is different when you define the function type. First, write the receiver object that you want to extend. Next, put a . and then complete the rest of your function type definition.

File: docs/topics/tour/kotlin-tour-intermediate-functions.md [hunk 2]
```diff
@@ -0,0 +1,935 @@
...
+that you want to extend followed by a `.` and the rest of your function type definition. For example: `MutableList<Int>.() -> Unit`.
+
+This function type has:
+
+* `MutableList<Int>` as the receiver type.
+* No function parameters within the parentheses `()`.
+* No return value: `Unit`.
+
+> Lambda expressions with receiver are also called function literals with receiver.
```
Comment (danil-pavlov) [inline docs/topics/tour/kotlin-tour-intermediate-functions.md:665]: This hint could be right after the 1st paragraph, I think here it's too late to add extra terms, in the middle of actual concept explanation

**Review by danil-pavlov — CHANGES_REQUESTED** [review]

**Comment by danil-pavlov** [comment]
Huge work! I think there's a lot of useful information and exercises but I feel like it's maybe too much for 1 page. I'm not sure what are the plans for the whole intermediate tour, but I'd split it into 3 separate steps, Extension functions, Scope functions, and Lambdas with receiver

File: docs/topics/tour/kotlin-tour-intermediate-functions.md [hunk 3]
```diff
@@ -0,0 +1,935 @@
...
+```
+{kotlin-runnable="true" id="kotlin-tour-scope-function-apply-before"}
+
+The example has a `Client` class that contains one property called `token` and three member functions: `connect()`,
+`authenticate()`, and `getData()`.
+
+The example creates `client` as an instance of the `Client` class before initializing its `token` property and calling its
+member functions in the `main()` function.
+
+Although this example is compact, in the real world it can be a while after the creation of your class instance before you
```
[the remark this answers is not in the document]
Comment (sarahhaggarty) [reply docs/topics/tour/kotlin-tour-intermediate-functions.md:290]: I'm trying to emphasize that it can be a while between the instance creation and the configuration of it. I think your suggested version makes that less clear so I'll keep the original :)

File: docs/topics/tour/kotlin-tour-intermediate-functions.md [hunk 4]
```diff
@@ -0,0 +1,935 @@
...
+{kotlin-runnable="true" id="kotlin-tour-scope-function-run"}
+
+The example:
+
+* Creates `client` as an instance of the `Client` class.
+* Uses the `.apply()` scope function on the `client` instance.
+* Creates a temporary scope within the `.apply()` scope function so that you don't have to explicitly refer to the `client` instance when accessing its properties or functions.
+* Passes a lambda expression to the `.apply()` scope function that updates the `token` property.
+
+|--|--|
```
[the remark this answers is not in the document]
Comment (sarahhaggarty) [reply docs/topics/tour/kotlin-tour-intermediate-functions.md:370]: I was trying to make it clear that the comments are related to before/after the main function and break it up so it's not just a wall of text. Do you think it would be better to combine the two lists?

File: docs/topics/tour/kotlin-tour-intermediate-functions.md [hunk 5]
```diff
@@ -0,0 +1,935 @@
...
+| `x.also { … }`  | `it`              | `x`           | Complete additional actions before returning the object.                                     |
+| `with(x) { … }` | `this`            | Lambda result | Call multiple functions on an object.                                                        |
+
+For more information about scope functions, see [Scope functions](scope-functions.md).
+
+## Scope functions practice
+
+### Exercise 1 {initial-collapse-state="collapsed" collapsible="true" id="scope-functions-exercise-1"}
+
+Rewrite the `getPriceInEuros` function as a single-expression function that uses Elvis operators `?.` and the `.let()` scope function.
```
[the remark this answers is not in the document]
Comment (sarahhaggarty) [reply docs/topics/tour/kotlin-tour-intermediate-functions.md:539]: Here, I've used the plural intentionally to try and give a hint to the reader that they should chain elvis operators. So I'll keep the original :)

File: docs/topics/tour/kotlin-tour-intermediate-functions.md [hunk 6]
```diff
@@ -0,0 +1,935 @@
+[//]: # (title: Intermediate functions)
+
+In this chapter, you will explore special Kotlin functions that make your code more concise and readable,
```
[the remark this answers is not in the document]
Comment (sarahhaggarty) [reply docs/topics/tour/kotlin-tour-intermediate-functions.md:3]: I think this comma helps to break up the sentence between "concise and readable" and "learn how they can help you". So unless you have a strong objection, I'd like to keep it as it is :)

[the remark this answers is not in the document]
Comment (danil-pavlov) [reply docs/topics/tour/kotlin-tour-intermediate-functions.md:3]: I just don't think it's grammatically necessary, but up to you

File: docs/topics/tour/kotlin-tour-intermediate-functions.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (danil-pavlov) [reply docs/topics/tour/kotlin-tour-intermediate-functions.md:370]: No, l'd vote for two separate lists with intro sentences for both.
because a single list seems too big, but the blank space between two lists seems like a bug

File: docs/topics/tour/kotlin-tour-intermediate-functions.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (danil-pavlov) [reply docs/topics/tour/kotlin-tour-intermediate-functions.md:290]: I get that, it's just the word order is a bit entangled for me. Another option:

> Although this example is compact, in the real world, it can take a while before you can configure and use the class instance (and its member functions) after you've created it.

**Review by danil-pavlov — CHANGES_REQUESTED** [review]

File: docs/topics/tour/kotlin-tour-intermediate-functions-part-2.md [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+
+In programming, a scope is the area in which your variable or object is recognized. The most commonly referred to scopes
+are global scope and local scope. In Kotlin, there are scope functions that allow you to create a temporary scope around
+an object and execute some code.
+
+Scope functions also make your code more concise because you don't have to refer to the name of your object within the temporary
+scope. Depending on the scope function, you can access the object either by referencing it via the keyword `this` or using it as an
+argument via the keyword `it`.
+
+Kotlin has five scope functions in total: `.let()`, `.apply()`, `.run()`, `.also()`, and `with()`.
```
[the remark this answers is not in the document]
Comment (sarahhaggarty) [reply docs/topics/tour/kotlin-tour-intermediate-functions-part-2.md:16]: Oh good catch! Thanks!
